client-websocket/src/main/java/com/github/twitch4j/client/websocket/WebsocketConnectionConfig.java

```diff
@@ -0,0 +1,121 @@
+package com.github.twitch4j.client.websocket;
+
+import com.github.twitch4j.common.config.ProxyConfig;
+import com.github.twitch4j.common.util.ExponentialBackoffStrategy;
+import com.github.twitch4j.util.IBackoffStrategy;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.experimental.Accessors;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.function.Consumer;
+
+@Accessors(chain = true, fluent = true, prefix = "")
+@Setter
+@Getter
+public class WebsocketConnectionConfig {
+
+    public static WebsocketConnectionConfig process(Consumer<WebsocketConnectionConfig> spec) {
+        WebsocketConnectionConfig data = new WebsocketConnectionConfig();
+        spec.accept(data);
+        data.validate();
+        return data;
+    }
+
+    /**
+     * validate the config
+     */
+    public void validate() {
+        Objects.requireNonNull(baseUrl, "baseUrl may not be null!");
+        if (wsPingPeriod < 0) {
+            throw new RuntimeException("wsPingPeriod must be 0 or greater, set to 0 to disable!");
+        }
+        Objects.requireNonNull(taskExecutor, "taskExecutor may not be null!");
+        Objects.requireNonNull(backoffStrategy, "backoffStrategy may not be null!");
+        Objects.requireNonNull(onPreConnect, "onPreConnect may not be null!");
+        Objects.requireNonNull(onPostConnect, "onPostConnect may not be null!");
+        Objects.requireNonNull(onConnected, "onConnected may not be null!");
+        Objects.requireNonNull(onTextMessage, "onTextMessage may not be null!");
+        Objects.requireNonNull(onTextMessage, "onTextMessage may not be null!");
+        Objects.requireNonNull(onDisconnecting, "onDisconnecting may not be null!");
+        Objects.requireNonNull(onPreDisconnect, "onPreDisconnect may not be null!");
+        Objects.requireNonNull(onPostDisconnect, "onPostDisconnect may not be null!");
+    }
+
+    /**
+     * The websocket url for the chat client to connect to.
+     */
+    private String baseUrl;
+
+    /**
+     * WebSocket RFC Ping Period in ms (0 = disabled)
+     */
+    private int wsPingPeriod = 0;
+
+    /**
+     * WebSocket Headers
+     */
+    private Map<String, String> headers = null;
+
+    /**
+     * Task Executor
+     */
+    private ScheduledExecutorService taskExecutor = new ScheduledThreadPoolExecutor(2);
+
+    /**
+     * Helper class to compute delays between connection retries
+     */
+    private IBackoffStrategy backoffStrategy = ExponentialBackoffStrategy.builder()
+        .immediateFirst(true)
```

Comment:
```suggestion
        .immediateFirst(false)
```
With the pubsub change below and connection pools relying on this, perhaps we delay the first reconnect to avoid a thundering herd issue